value-processor/src/org/immutables/value/processor/meta/AttributeBuilderReflection.java

```diff
@@ -0,0 +1,463 @@
+      throw new AssertionError();
+    }
+
+    String cacheKey = cachingKey(valueAttribute());
+    if (analyzedReturnTypes.containsKey(cacheKey)) {
+      return Preconditions.checkNotNull(analyzedReturnTypes
+          .get(cacheKey));
+    }
+
+    AttributeBuilderDescriptor descriptor = getReflectionStrategy().getAttributeBuilderDescriptor();
+    analyzedReturnTypes.put(cacheKey, descriptor);
+
+    return descriptor;
+  }
+
+  @Lazy
+  protected Strategy getReflectionStrategy() {
+    for (Strategy strategy : getStrategies()) {
+      if (strategy.isAttributeBuilder()) {
+        return strategy;
+      }
+    }
+
+    throw new AssertionError("isAttributeBuilder flip-flopped from true to false.");
+  }
+
+  /**
+   * Allows for different mechanisms of discovering nested builders.
+   *
+   * This is needed because Immutables value objects may not be available on the
+   * class path during processing, which will be order dependent.
+   */
+  interface Strategy {
+
+    boolean isAttributeBuilder();
+
+    AttributeBuilderDescriptor getAttributeBuilderDescriptor();
+  }
+
+  /**
+   * Strategy for processing first party immutables.  Honors both deepImmutableDiscovery, and
+   * builder extension.
+   */
+  @Immutable(builder = false)
+  abstract static class FirstPartyStrategy implements Strategy {
+
+    @Parameter
+    abstract ValueAttribute valueAttribute();
+
+    @Override
+    public boolean isAttributeBuilder() {
+      return valueAttribute().attributeValueType != null &&
+          valueAttribute().attributeValueType.isUseBuilder();
+
+    }
+
+    @Override
+    public AttributeBuilderDescriptor getAttributeBuilderDescriptor() {
+      return ImmutableAttributeBuilderDescriptor.builder()
+          .valueToBuilderTarget(ValueToBuilderTarget.BUILDER_INSTANCE)
+          .valueToBuilderMethod(attributeValueType().names().from)
+          .buildMethod(attributeValueType().names().build)
+          .qualifiedValueTypeName(attributeValueType().typeImmutable().toString())
+          .qualifiedBuilderTypeName(attributeValueType().typeBuilderImpl().toString())
+          .qualifiedBuilderConstructorMethod(attributeValueType().factoryBuilder().toString())
+          .build();
+
+    }
+
+    ValueType attributeValueType() {
+      return valueAttribute().attributeValueType;
+    }
+  }
+
+  /**
+   * Strategy for parsing third party immutables. for example: the protocol buffer API.
+   *
+   * Assumes that all third party classes are available on the class path at processing time.
+   *
+   * If this is not the case, we may need to use a processing method that allows deferring of
+   * compilation. Example, moving from auto-value to immutables... though maybe you would just
+   * implement a new strategy for that use case...
+   */
+  @Immutable(builder = false)
+  abstract static class ThirdPartyStaticBuilderStrategy implements Strategy {
+
+    private ExecutableElement buildMethod = null;
```

Comment:
grumble... so this isn't an immutable anymore... not sure the best way to clean this up unfortunately. It's a lot cleaner this way than before when I had a bunch of lazy optionals.